titiler/pgstac/factory.py

```diff
@@ -861,35 +390,35 @@ def wmts(  # noqa: C901
                 "use_epsg",
                 "request",
             ]
-            qs = [
-                (key, value)
-                for (key, value) in request.query_params._list
-                if key.lower() not in qs_key_to_remove
-            ]
+            qs = urlencode(
+                [
+                    (key, value)
+                    for (key, value) in request.query_params._list
+                    if key.lower() not in qs_key_to_remove
+                ],
+                doseq=True,
+            )
 
             # Checking if we can construct a valid tile URL
             # 1. we use `check_query_params` to validate the query-parameter
             # 2. if there is no layers (from mosaic metadata) we raise the caught error
             # 3. if there no errors we then add a default `layer` to the layers stack
-            try:
-                check_query_params(
-                    dependencies=tile_dependencies,  # type: ignore
-                    query_params=QueryParams(qs),
-                )
-            except Exception as e:
-                if not layers:
-                    raise e
-            else:
+            if check_query_params(tile_dependencies, QueryParams(qs)):
                 layers.append(
                     {
                         "titler": search_info.metadata.name or search_id,
                         "name": "default",
                         "tiles_url": tiles_url,
-                        "query_string": urlencode(qs, doseq=True) if qs else None,
+                        "query_string": qs if qs else None,
                         "bounds": bounds,
                     }
                 )
 
+            if not layers:
+                raise NoLayerFound(
+                    "Could not find any valid layers in metadata or construct one from Query Parameters."
+                )
+
```

Comment:
we don't raise message from `check_query_params` function so we just raise a simple error if there are no layers